### ISSUE TITLE
Rewrite app probe sends requests to the wrong endpoint, breaking pod IP listeners

### ISSUE THREAD

**Issue by vvavepacket** [istio/istio#25177]
**Bug description**
When kubernetes is probing the pod's readiness / liveness probe, it is always returning connection refused
```
2020-07-02T03:16:51.019635Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
2020-07-02T03:16:57.829163Z error Request to probe app failed: Get "http://localhost:8558/alive": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/livez
app URL path = /alive
2020-07-02T03:17:01.020098Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
2020-07-02T03:17:07.832858Z error Request to probe app failed: Get "http://localhost:8558/alive": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/livez
app URL path = /alive
2020-07-02T03:17:11.019798Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
2020-07-02T03:17:17.828587Z error Request to probe app failed: Get "http://localhost:8558/alive": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/livez
app URL path = /alive
2020-07-02T03:17:21.019642Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
2020-07-02T03:17:27.829042Z error Request to probe app failed: Get "http://localhost:8558/alive": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/livez
app URL path = /alive
2020-07-02T03:17:31.019802Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
2020-07-02T03:17:51.019767Z error Request to probe app failed: Get "http://localhost:8558/ready": dial tcp
127.0.0.1:8558: connect: connection refused, original URL path = /app-health/miniapp/readyz
app URL path = /ready
```
Apparently this [PR ](https://github.com/istio/istio/pull/13970) should have fixed this but I tested on this version and its failing

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Readiness probe and liveness probe should be accessible. I am able to access the probes directly from other pods by curl and its returning 200 OK. It seems probe request are still going thru the sidecar proxy
```
curl http://10.244.1.231:8558/ready
OK
```

**Steps to reproduce the bug**
Here is my sample pod configuration (as we can see I have added the probes to outbound exclusion port so it doesn't pass thru envoy sidecar proxy, also only 9000 traffic will pass thru into the sidecar proxy so traffic going into probes listening on port 2552, 8558 will be accessible directly without going thru the sidecar proxy)

1. Install istio
2. Deploy pod to namespace with injection enabled
3. As pod is starting, it will probe checks will fail

Here is my pod yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: miniapp-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      component: miniapp
      app: miniapp-lagom
  template:
    metadata:
      labels:
        component: miniapp
        app: miniapp-lagom
      annotations:
        traffic.sidecar.istio.io/includeInboundPorts: "9000"
        traffic.sidecar.istio.io/excludeOutboundPorts: "2552,8558"
        
    spec:
      containers:
        - name: miniapp
          #image: 192.168.1.3:5000/miniapp-impl:0.1.2
          imagePullPolicy: Always
          env:
            - name: REQUIRED_CONTACT_POINT_NR
              value: "1"
            - name: JAVA_OPTS
              value: "-Dconfig.resource=application-k8s.conf"
            - name: PGPASSWORD
              valueFrom:
                secretKeyRef:
                  name: miniapp 
                  key: pgpassword
          ports:
            - name: management
              containerPort: 8558 
              protocol: TCP
            - name: live
              containerPort: 9000
          readinessProbe:
            httpGet:
              path: "/ready"
              port: management
            periodSeconds: 10
            failureThreshold: 10
            initialDelaySeconds: 20
            timeoutSeconds: 1
          livenessProbe:
            httpGet:
              path: "/alive"
              port: management
            periodSeconds: 10
            failureThreshold: 10
            initialDelaySeconds: 20
            timeoutSeconds: 1
      imagePullSecrets:
        - name: awsecr-cred
```
Here is my service yaml
```
apiVersion: v1
kind: Service
metadata:
  name: miniapp-cluster-ip-service
spec:
  type: NodePort
  selector:
    component: miniapp
    app: miniapp-lagom
  ports:
    - port: 3101
      targetPort: 9000
      name: http-live
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
istioctl : 1.6
kubectl : 1.17
**How was Istio installed?**
I tried two configurations: minimal and demo and issue persists on both
istioctl manifest apply --set profile=minimal --set values.sidecarInjectorWebhook.rewriteAppHTTPProbe=true
istioctl manifest apply --set profile=demo --set values.sidecarInjectorWebhook.rewriteAppHTTPProbe=true

**Environment where bug was observed (cloud vendor, OS, etc)**
Bare metal kubernetes cluster running on top of Ubuntu

I'm getting below error when attempting to extract cluster state archive
`bash: ./dump_kubernetes.sh: Permission denied`

**Comment by howardjohn** [comment]
Does /ready listen on the pod ip? maybe that is why request to localhost fails. I don't think the problem is going through envoy

**Comment by vvavepacket** [comment]
@howardjohn 
Hi, I think yes its listening to the pod IP, I tried getting the pod's IP and curling it directly and it results into Success. Sorrty I didnt get why the request to localhost fails, where should the /ready be listening into instead?

**Comment by howardjohn** [comment]
0.0.0.0 or 127.0.0.1

@incfly is this expected that we require this? shouldn't we send to the same destination that kubelet does?

**Comment by howardjohn** [comment]
"The kubelet sends the probe to the pod’s IP address, unless the address is overridden by the optional host field in httpGet."

I think we need this logic in the rewrite probe if we don't already

**Comment by howardjohn** [comment]
If we change to sending to pod IP, we need to be careful about what happens if existing users are listening on localhost. We may need a flag or big release note

**Comment by howardjohn** [comment]
cc @sdake

**Comment by howardjohn** [comment]
not stale

**Comment by howardjohn** [comment]
This is possibly going to be addressed in 1.10. working on a design

**Comment by howardjohn** [comment]
Fixed by #31121
